### PR TITLE
Add filters to Allocations

### PR DESCRIPTION
**Pull request by ChaiWithJai**
Mock
image

Actual
image

### CODE REVIEW

File: ui/app/controllers/jobs/job/allocations.js [hunk 1]
```diff
@@ -40,11 +57,32 @@ export default class AllocationsController extends Controller.extend(
...
+  @computed('model.allocations.[]', 'selectionStatus', 'selectionClient', 'selectionTaskGroup')
   get allocations() {
-    return this.get('model.allocations') || [];
+    const allocations = this.get('model.allocations') || [];
+    const { selectionStatus, selectionClient, selectionTaskGroup } = this;
+
+    if (!allocations.length) return allocations;
+
+    return allocations.filter(alloc => {
+      if (selectionStatus.length && !selectionStatus.includes(alloc.status)) {
```
Comment (lgfa29) [inline ui/app/controllers/jobs/job/allocations.js:68]:
```suggestion
      if (selectionStatus.length && !selectionStatus.includes(alloc.clientStatus)) {
```

File: ui/app/controllers/jobs/job/allocations.js [hunk 2]
```diff
@@ -53,4 +91,45 @@ export default class AllocationsController extends Controller.extend(
...
+  get optionsClients() {
+    const clients = Array.from(new Set(this.model.allocations.mapBy('node.shortId'))).compact();
+
+    // Update query param when the list of clients changes.
+    scheduleOnce('actions', () => {
+      // eslint-disable-next-line ember/no-side-effects
+      this.set('qpClient', serialize(intersection(clients, this.selectionClient)));
+    });
+
+    return clients.sort().map(dc => ({ key: dc, label: dc }));
```
Comment (lgfa29) [inline ui/app/controllers/jobs/job/allocations.js:116]:
```suggestion
    return clients.sort().map(c => ({ key: c, label: c }));
```

File: ui/app/controllers/jobs/job/allocations.js [hunk 3]
```diff
@@ -53,4 +91,45 @@ export default class AllocationsController extends Controller.extend(
...
+  get optionsTaskGroups() {
+    const taskGroups = Array.from(new Set(this.model.allocations.mapBy('taskGroupName'))).compact();
+
+    // Update query param when the list of clients changes.
+    scheduleOnce('actions', () => {
+      // eslint-disable-next-line ember/no-side-effects
+      this.set('qpTaskGroup', serialize(intersection(taskGroups, this.selectionTaskGroup)));
+    });
+
+    return taskGroups.sort().map(dc => ({ key: dc, label: dc }));
```
Comment (lgfa29) [inline ui/app/controllers/jobs/job/allocations.js:129]:
```suggestion
    return taskGroups.sort().map(tg => ({ key: tg, label: tg }));
```

**Review by lgfa29 — CHANGES_REQUESTED** [review]
Looking good! 

Filtering by status doesn't work because the filter is reading the wrong alloc property (`status` vs `clientStatus`). It would be good to have some tests for this as well, they would've likely caught this problem. Take a look at other pages with filters to see how they are done.

I think filtering by job version would could be helpful as well, and seems like a quick addition. But feel free to drop it if it's not.

File: ui/app/controllers/jobs/job/allocations.js [hunk 4]
```diff
@@ -1,17 +1,22 @@
...
+import { serialize, deserializedQueryParam as selection } from 'nomad-ui/utils/qp-serialize';
 import classic from 'ember-classic-decorator';
 
 @classic
 export default class AllocationsController extends Controller.extend(
-    Sortable,
-    Searchable,
-    WithNamespaceResetting
-  ) {
+  /*eslint-disable  indent */
```
[the remark this answers is not in the document]
Comment (ChaiWithJai) [reply ui/app/controllers/jobs/job/allocations.js:15]: I can't fix it.

File: ui/app/controllers/jobs/job/allocations.js [hunk 5]
```diff
@@ -53,4 +104,58 @@ export default class AllocationsController extends Controller.extend(
...
+      { key: 'running', label: 'Running' },
+      { key: 'complete', label: 'Complete' },
+      { key: 'failed', label: 'Failed' },
+      { key: 'lost', label: 'Lost' },
+    ];
+  }
+
+  @computed('model.allocations.[]', 'selectionClient')
+  get optionsClients() {
+    const clients = Array.from(new Set(this.model.allocations.mapBy('node.shortId'))).compact();
```
Comment (LevelbossMike) [inline ui/app/controllers/jobs/job/allocations.js:121]: We are repeating this multiple times in this file. I understand that this pattern is used in other places as well `controllers/clients/index.js` for example. It looks like we can create a computed-macro out of this pattern.

File: ui/app/controllers/jobs/job/allocations.js [hunk 6]
```diff
@@ -53,4 +104,58 @@ export default class AllocationsController extends Controller.extend(
...
+
+    // Update query param when the list of clients changes.
+    scheduleOnce('actions', () => {
+      // eslint-disable-next-line ember/no-side-effects
+      this.set('qpClient', serialize(intersection(clients, this.selectionClient)));
```
Comment (LevelbossMike) [inline ui/app/controllers/jobs/job/allocations.js:126]: we probably don't want to trigger side-effects in computeds. Again this is done in other places as well but we should investigate why we feel this is necessary to do

File: ui/app/controllers/jobs/job/allocations.js [hunk 7]
```diff
@@ -40,11 +61,41 @@ export default class AllocationsController extends Controller.extend(
...
+    'selectionClient',
+    'selectionTaskGroup',
+    'selectionJobVersion'
+  )
   get allocations() {
-    return this.get('model.allocations') || [];
+    const allocations = this.get('model.allocations') || [];
+    const { selectionStatus, selectionClient, selectionTaskGroup, selectionJobVersion } = this;
+
+    if (!allocations.length) return allocations;
```
Comment (LevelbossMike) [inline ui/app/controllers/jobs/job/allocations.js:75]: we can just work with `model.allocation` directly here probably - i.e. directly go to the filter.

**Review by LevelbossMike — APPROVED** [review]
We can tackle the macro in a follow-up PR and apply it to the rest of the codebase that also uses this pattern.

Good work! 👏

**Review by lgfa29 — APPROVED** [review]
> I think filtering by job version would could be helpful as well

On second thought, it wasn't very helpful. Long-lived jobs can have their version in the hundreds (if not thousands), the filter list would be pretty useless and hard to use. 

So I removed this filter. Sorry for the unnecessary work 😅

**Pull request merged**